### DISCUSSION TITLE
Remote Rendering with high FPS, Browser doesn't show intermediate frames, just the last one

### DISCUSSION THREAD

**Discussion by jcorbettfrank** [Kitware/trame#302] [Q&A] [answered]
Hi, 

Is the behavior below expected? When the desired frame rate on the wavelets.py async-paraview example is set too high, the browser receives images but does not display them; it only displays the last one and after the interaction is over. 

https://github.com/Kitware/trame/assets/28637033/87b58d8d-d32b-41a9-9cf1-bb4e7e57239b

I ask here because this same behavior is seen with our own application  (vtkWebApplication [very similar to https://github.com/Kitware/trame-vtk/tree/master/trame_vtk/modules/vtk/protocols]  + Angular + wslink), specifically when scrolling. Our thought is that there is no rate-limiter on the scrolls and hence no limit on images sent from the server to client. Even if the browser can only render at a certain rate, shouldn't some intermediate frames be rendered, instead of just the last one, as seen in the video?

Is this behavior expected? Is this a bug? If this should be posted elsewhere please let me know. Thanks for your time.

**Comment by jourdain** [comment, marked as answer]
This could be expected as the decoder pool could be too small on the client side. (I think we can increase it but probably hardcoded currently). Also if we make it too big, we'll introduce some lag buffer which is not great either. 
Basically we are overflowing it in a way that things get canceled before any image get resolved. 

In general, you don't want to send more data than the client can decode. So ideally you should instrument the client side to adjust the server rate. So far we are exposing the knobs but not the automation to tune them up.  

On the standard ParaView (web) side of things we never were going beyond 30 fps but also because of the synchronous architecture we were lowering that target based on the busyness of the main paraview thread.
While on the async side, that main thread is 99% free so we can't rely on the server side anymore. The client needs to start sharing its state to drive a reasonable frame rate.

Hope that make sense,

Seb

**Reply by jcorbettfrank** [reply]
Thanks for your response.

Yes, that makes sense. If the client can tell it is filling the decoding pool faster than images can be rendered on the browser, it can send a set.animation.fps(lower) call (paraphrasing). For scrolling, it may be a bit different because it's discrete, but we can still implement some rate limiter whether that comes from the client's sending of mouse.zoom events or some playback widget with a play/pause button etc.

Thank you again, I appreciate the detailed response!

Jordan
